### ISSUE TITLE
versions:2.17.0:set updates grandchilds with wrong version

### ISSUE THREAD

**Issue by schwarfl** [mojohaus/versions#1103]
Hi,

I think there is a bug in versions:2.17.0:set when used in a maven reactor with different versions. We have 

> 
```
Grandparent: 
  <artifactId>Grandparent</artifactId>
  <version>1</version>
 
Parent:
   <parent>
     <artifactId>Grandparent</artifactId>
     <version>1</version>
  </parent>
  <artifactId>Parent</artifactId>
  <version>100-SNAPSHOT</version>
 
Child:
   <parent>
     <artifactId>Parent</artifactId>
     <version>100-SNAPSHOT</version>
  </parent>
  <artifactId>Child</artifactId>

Grandchild:
   <parent>
     <artifactId>Child</artifactId>
     <version>100-SNAPSHOT</version>
  </parent>
  <artifactId>Grandchild</artifactId>
``` 

If one calls
mvn versions:2.17.0:set -DnewVersion=2
Not only the version in grandparent and parent gets updated from 1 to 2 but also the version in Grandchild gets updated from 100-SNAPSHOT to 2. Also the log messages contains
[INFO] Processing change of com.company.test:grandparent:1 -> 2
[INFO] Processing com.company.test:grandparent
[INFO]     Updating project com.company.test:grandparent
[INFO]         from version 1 to 2
[INFO]
[INFO] Processing com.company.test:parent
[INFO]     Updating parent com.company.test:grandparent
[INFO]         from version 1 to 2
[INFO]
[INFO] Processing com.company.test:grandchild
[INFO]     Updating parent com.company.test:child
[INFO]         from version 100-SNAPSHOT to 2

If one calls
mvn versions:2.16.2:set -DnewVersion=2
instead, only grandparent and parent gets updated from 1 to 2.

Attached you find a minimal set to reproduce the error
[testVersionsSet.zip](https://github.com/user-attachments/files/16049670/testVersionsSet.zip)

**Comment by svetlana-nikitina** [comment]
we experience a similar issue with `-pl` flag:
```
 mvn versions:set -DnewVersion=1.10.0 -pl module-name
 ```
 updates versions across the whole project and not only one single module

**Comment by mricken** [comment]
We also noticed the behavior that @svetlana-nikitina mentioned. Updating in a single module updates across the entire project. Version `2.16.2` didn't do that.

**Comment by slawekjaranowski** [comment]
Thanks for reporting .... I would like to start to check again change: #1051

We can start by add attached example as IT

I can try to work on it but in next week ... and end of July I will have 🌴 
So any help can be valuable

**Comment by slawekjaranowski** [comment]
> we experience a similar issue with `-pl` flag after the new release:
> 
> ```
>  mvn versions:set -DnewVersion=1.0.0 -pl module-name
> ```
> 
> updates versions across the whole project and not only one single module

next example which can be ease added as IT.

**Comment by slawekjaranowski** [comment]
Does any one want to help .... I will create an issue #1106 and first part of fix in #1107
There is many other ITs for cleanups. 

It will be easier to understand where is a bug.
